src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -562,19 +590,21 @@ private void RecordWaiterStartTime()
                 // Don't record zero, that value is reserved for indicating that a time is not recorded
                 currentTimeMs--;
             }
-            _waiterStartTimeMs = currentTimeMs;
+            WaiterStartTimeMs = currentTimeMs;
         }
 
         private bool ShouldStopPreemptingWaiters
         {
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
             get
             {
-                // If the recorded time is zero, a time has not been recorded yet
-                ushort waiterStartTimeMs = _waiterStartTimeMs;
+                // If the recorded time is zero, a time has not been recorded yet. The stored waiter start time (ms as a ushort)
+                // excludes the upper bit since the field is also used for a flag, so also mask off the upper bits from the
+                // current tick count for comparison.
+                ushort waiterStartTimeMs = WaiterStartTimeMs;
                 return
                     waiterStartTimeMs != 0 &&
-                    (ushort)Environment.TickCount - waiterStartTimeMs >= MaxDurationMsForPreemptingWaiters;
+                    (Environment.TickCount & 0x7fff) - waiterStartTimeMs >= MaxDurationMsForPreemptingWaiters;
```

Comment:
Thanks, fixed. I think I the `& 0x7fff` is needed on the computed duration, since the recorded start time excludes the upper bit of a ushort. For instance, if the actual start time was `0xffff`, `0x7fff` would be recorded, and if the current time is `0x10000`, the diff would be 0x8001 ms instead of 1 ms.